Editorial - Alchemy (#6381)
cpinitiative/usaco-guide: solutions/bronze/usaco-1229.mdx

---
id: usaco-1229
source: USACO Bronze 2022 US Open
title: Alchemy
author: Rameez Parwez
---

[Official Editorial (C++, Java)](https://usaco.org/current/data/sol_prob3_bronze_open22.html)

## Explanation

Bessie the cow wants to maximise her units of metal $N$, for which
she can use the recipes provided to convert other constituent
metals into metal $N$.

We can recursively check the $N$-th metal constituents, and if it
has enough availability we will consume one unit of it and check
for other metal constituents. We will repeat this until we run out

> of supply for constituent metals.

Reviewer: should address why this greedy works:
- since there's only one recipe for each metal, it's a binary choice between constructing a unit of the metal, and using up a unit that already exists
- we can always delay constructing a new unit of metal, so we should prioritize using what we already have and constructing what's needed later if needed
- recursion is easiest way to impl

Author: okay, thanks for the points 👍

## Implementation

> **Time Complexity:** $$

Reviewer: add time comp?

ik editorial states code is $\mathcal{O}(N^2 A)$ but recursive code is probably slower cuz ur revisiting nodes a lot more (eg consider a complete graph)

Author: is it somethinh like $O(2^{\text{max depth of recipe chain}})$, if i am not tripping

Reviewer: probably? i'm not 100% sure to be honest

you can also mention the iterative approach when noting the time complexity

Author: > you can also mention the iterative approach when noting the time complexity

I don't get it

Reviewer: since I think the time complexities of the iterative and recursive approaches are different, you can note that the iterative approach is faster / has a different time complexity

(correct me if im wrong tho)

Author: yup, u r right

<Info title = "Note">

The intuitive approach of this problem is faster than this 
recursive approach. Can you figure out what it might be?

</Info>

will this suffice?

<LanguageSection>
<CPPSection>

```cpp
#include <functional>
#include <iostream>
#include <unordered_map>
#include <vector>

int main() {
	int n;
	std::cin >> n;
	std::vector<int> metals(n);
	for (int &x : metals) { std::cin >> x; }

	int k;
	std::cin >> k;

> 	std::unordered_map<int, std::vector<int>> recipes;

Reviewer: why not `std::vector<std::vector<int>> recipes(n)` instead

cuz $l$ is the index for a metal which is bounded by $n$

Author: just preference, not reason in particular.

should i change map -> vector

Reviewer: generally yes if it's convenient for the problem

Author: okay

---
id: usaco-1229
source: USACO Bronze 2022 US Open
title: Alchemy
author: Rameez Parwez
---

> 
> [Official Editorial (C++, Java)](https://usaco.org/current/data/sol_prob3_bronze_open22.html)
> 
> ## Explanation
> 
> Bessie the cow wants to maximise her units of metal $N$, for which
> she can use the recipes provided to convert other constituent
> metals into metal $N$.
> 
> We can recursively check the $N$-th metal constituents, and if it
> has enough availability we will consume one unit of it and check
> for other metal constituents. We will repeat this until we run out
> of supply for constituent metals.
> 
> This recursive greedy approach works perfectly for this problem for
> the following reasons:
> 
> 1. Since there is only one recipe for each metal, constructing a
> new unit of any metal is a binary choice: either we consume from
> our existing stock, or we construct a new unit from its recipe.
> 
> 2. We can always delay constructing a new unit of a metal, so we
> should prioritize using our existing stock first and build new
> units of metal only if necessary.
> 
> 3. The ordering of recipes forms a tree-like structure, so a
> recursive approach is easier to implement than other approaches.

Reviewer: ```suggestion
Bessie the cow wants to maximise her units of metal $N$, for which
she can use the recipes provided to convert other constituent
metals into metal $N$.

We can recursively check the $N$-th metal constituents, and if it
has enough availability, we will consume one unit of it and check
for other metal constituents. We will repeat this until we run out
of supply for constituent metals.

We recursively check if it is possible to create a unit of metal $N$.
For the current metal, we either consume
an existing unit of it, or we recursively construct the necessary metals in its recipe. 
We construct units of metal $N$ recursively until it is no longer possible.

This recursive greedy approach works perfectly for this problem for
the following reasons:

- Since there is only one recipe for each metal, constructing a
new unit of any metal is a binary choice: either we consume from
our existing stock, or we construct a new unit from its recipe.

- We can always delay constructing a new unit of metal, so we
should prioritize using our existing stock first and build new
units of metal only if necessary.

The ordering of recipes forms a tree-like structure, so a
recursive approach is easier to implement than other approaches.
```

## Implementation

**Time Complexity:** $\mathcal{O}(2^{max\_depth\_of\_recipe\_chain})$

<Info title = "Note">

The intuitive approach of this problem is faster than this

> recursive approach. Can you figure out what it might be?

Reviewer: ```suggestion
The iterative implementation of this algorithm is more efficient. As an exercise,
try implementing an iterative approach that runs in $\mathcal{O}(N^2 \cdot \max(a_i), using the fact
that recipes utilize strictly less valuable metals.
```

> </Info>

Reviewer: ```suggestion
**Time Complexity:** $\mathcal{O}(N \cdot \sum a_i)$

<Info>

At first glance, the recursive code appears to be exponential. However, this is not the case due to amortization.

Consider an individual unit of a metal $i$. If possible, it will eventually be used to create a unit of metal $N$, meaning that it will contribute to at most $N$ calls. Since our source of metal can only deplete, we can analyze the individual contributions of each unit of metal to analyze the complexity of the algorithm.

</Info>
```

i believe the reasoning is correct ? since during the time it's possible to create a unit of metal $N$, the logic holds, and because we terminate early if it's not possible, idt there'd be any time complexity blow-ups
